### PR TITLE
[Sema] Remove separate closure type-checking logic

### PR DESCRIPTION
**Pull request by hamishknight**
`participatesInInference` is now always true for a non-empty body, remove it along with the separate type-checking logic such that empty bodies are type-checked together with the context.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

File: lib/Sema/CSDiagnostics.cpp [hunk 1]
```diff
@@ -5515,6 +5533,55 @@ bool MissingArgumentsFailure::diagnoseClosure(const ClosureExpr *closure) {
...
+  auto *UDE = dyn_cast<UnresolvedDotExpr>(call->getFn());
+  if (!UDE || !UDE->isImplicit())
+    return false;
+
+  auto overload = getCalleeOverloadChoiceIfAvailable(getLocator());
+  if (!overload)
+    return false;
+
+  auto *decl = overload->choice.getDeclOrNull();
+  if (!decl || decl->getBaseName() != ctx.Id_buildBlock)
```
Comment (xedin) [inline lib/Sema/CSDiagnostics.cpp:5558]: There is a method we use already to determine that based on expression itself - `isResultBuilderMethodReference` all you'd need to do is check whether it's also `Id_buildBlock` specifically as well. 

For completeness you might also want to cover `Id_buildPartialBlock` I think.

**Review by xedin — APPROVED** [review]

**Comment by hamishknight** [comment]
@swift-ci please test

File: lib/Sema/CSDiagnostics.cpp (hunk 1 above)
Comment (hamishknight) [reply lib/Sema/CSDiagnostics.cpp:5557]: Ah nice, updated to use `isResultBuilderMethodReference`. `buildPartialBlock` is only used for non-empty cases I believe, so isn't applicable at least to the current case being handled here.

**Comment by hamishknight** [comment]
@swift-ci please smoke test macOS

**Pull request merged**